File: LeetCode/264/main.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        if(n == 1) {
            return 1;
        }
        int p1 = 0, p2 = 0, p3 = 0;
        vector<int> res(n, 0);
        res[0] = 1;

        // 3 pointers are updated on the res array as 
        // we want the last pos : the last element where 
        // that factor was used)
        for(int i = 1; i < n; i++) {
            int t1 = 2 * res[p1];
            int t2 = 3 * res[p2];
            int t3 = 5 * res[p3];
            
            int t4 = min({t1, t2, t3});
            if(t4 == t1) {
                p1++;
            } 
            if(t4 == t2) {
                p2++;
            } 
            if(t4 == t3) {
                p3++;
            }

            res[i] = t4;
        }
        
        return res[n-1];
    }
};

File: LeetCode/264/main.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:

        res = [0] * n
        res[0] = 1

        p1, p2, p3 = 0, 0, 0

        for i in range(1, n, 1):

            t1, t2, t3 = 2 * res[p1], 3 * res[p2], 5 * res[p3]
            t4 = min(t1, t2, t3)

            if (t4 == t1):
                p1 += 1
            if (t4 == t2):
                p2 += 1
            if (t4 == t3):
                p3 += 1

            res[i] = t4

        return res[n - 1]
